### PR TITLE
Issue215

### PR DESCRIPTION
**Pull request by Flumble**
This pull request resolves issue #215

### CODE REVIEW

**Comment by wandernauta** [comment]
Thanks, I'll take a look.

File: Code/src/main/java/nl/utwente/viskell/ui/components/SliderBlock.java [hunk 1]
```diff
@@ -11,14 +11,14 @@
...
+    @FXML protected Slider slider;
 
     /**
      * Constructs a new SliderBlock
      * @param pane The parent pane this Block resides on.
      */
-    public SliderBlock(CustomUIPane pane) {
-        super(pane, Type.con("Float"), "0.0", "SliderBlock");
+    public SliderBlock(CustomUIPane pane, Environment env) {
+        super(pane, env.buildType("Fractional a => a"), "0.0", "SliderBlock");
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/viskell/ui/components/SliderBlock.java]: Why pass `env` as an argument when you can just as easily do `pane.getEnvInstance` here?

File: Code/src/main/java/nl/utwente/viskell/ui/components/ValueBlock.java [hunk 2]
```diff
@@ -22,16 +24,17 @@
  */
 public class ValueBlock extends Block {
     /** The value of this ValueBlock. */
-    private StringProperty value;
+    protected StringProperty value;
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/viskell/ui/components/ValueBlock.java]: There are public `(get|set)value(property)` methods available. You could just use those from ValueBlock's subclasses.

File: Code/src/main/java/nl/utwente/viskell/ui/components/ValueBlock.java [hunk 3]
```diff
@@ -52,6 +55,8 @@ protected ValueBlock(CustomUIPane pane, Type type, String value, String fxml) {
 
         outputSpace.setCenter(this.getOutputAnchor().get());
         outputSpace.toFront();
+        
+        this.valueType.setText("hoi");
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/viskell/ui/components/ValueBlock.java]: Seems like debug code.

File: Code/src/main/java/nl/utwente/viskell/ui/components/ValueBlock.java [hunk 4]
```diff
@@ -22,16 +24,17 @@
...
+    protected StringProperty value;
 
     /** The OutputAnchor of this ValueBlock. */
-    private OutputAnchor output;
+    protected OutputAnchor output;
 
     /** The space containing the output anchor. */
-    @FXML private BorderPane outputSpace;
+    @FXML protected BorderPane outputSpace;
+    @FXML protected Label valueType;
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/viskell/ui/components/ValueBlock.java]: Missing Javadoc for `valueType` makes it seem it relates to the Javadoc for `outputSpace`.

File: Code/src/main/java/nl/utwente/viskell/ui/components/ValueBlock.java [hunk 5]
```diff
@@ -88,6 +93,16 @@ public void updateExpr() {
...
+    
+    @Override
+    public void invalidateVisualState() {
+        try {
+			this.valueType.setText(this.expr.findType().prettyPrint());
+		} catch (HaskellTypeError e) {
+			this.valueType.setText("???");
+		}
+    	super.invalidateVisualState();
+    }
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/viskell/ui/components/ValueBlock.java]: We use four-space indents, no tabs.

File: Code/src/main/resources/ui/style.css [hunk 6]
```diff
@@ -65,14 +65,21 @@ CustomUIPane {
     -fx-text-alignment: center;
 }
 
-.display.block .argumentType {
-    -fx-padding: 0 5 0 5;
+.display.block .argumentType, .value.block .argumentType {
```
Comment (wandernauta) [inline Code/src/main/resources/ui/style.css]: I'd say this is not an argument type (black-on-white) but a result type (white-on-black).

**Comment by Flumble** [comment]
The colour of the valueblock types is reverted to black-on-white again, because the inverse is horrible, despite being more correct.

I believe subclasses (_SliderBlock_) should have more direct access to their properties (_ValueBlock::StringProperty_) than other classes _by default_. Right now the public getter and setter are simply returning the property as-is, but restrictions may be added in the future that shouldn't affect subclasses.

**Comment by wandernauta** [comment]
If the inverse is horrible, why not change it elsewhere, as well? You've now introduced an inconsistency which would have to be explained ('white is in, black is out, except on such-and-such blocks').

As for the `protected` thing, I'd of course argue the contrary - subclasses should have as little access to *anything* as possible. In this case, SliderBlock should not depend on the fact that ValueBlock happens to have a property called 'value'. It should instead call ValueBlock's public API, which it seems it already does; ValueBlock should then be free to implement `getValue` and `setValue` in any way it sees fit. [ref](http://stackoverflow.com/a/2279692/182402)

**Comment by comonoidial** [comment]
As for the background of types let's just make them all white for now, and figure out better visuals at a later stage.
I find it unproductive to discuss something minor as private vs protected in package being rewritten.

**Pull request merged**